packages/utils/src/groupInMap.ts

```diff
@@ -18,35 +10,23 @@ const DELETE = Symbol("DELETE")
  * @param keyGetter A function that extracts the key for each item.
  * @param projection Projection function for each group.
  */
-export const groupInMap = <T, K, V>(
-  keyGetter: (x: T) => K,
-  projection: (x: GroupedObservable<K, T>) => Observable<V>,
-) => (source$: Observable<T>): Observable<Map<K, V>> => {
+export const groupInMap = <K, V>() => (
+  source$: Observable<GroupedObservable<K, V>>,
+): Observable<Map<K, V>> => {
   const res = new Map<K, V>()
 
-  return concat(
-    source$.pipe(
-      continuousGroupBy(keyGetter),
-      publish((multicasted$) => {
-        return multicasted$.pipe(
-          mergeMap((inner$) =>
-            concat(
-              projection(inner$).pipe(map((v) => [inner$.key, v] as const)),
-              of([inner$.key, DELETE] as const),
-            ),
-          ),
-          takeUntil(multicasted$.pipe(takeLast(1))),
-        )
-      }),
-      scan((acc, [key, value]) => {
-        if (value !== DELETE) return acc.set(key, value)
-        acc.delete(key)
-        return acc
-      }, res),
+  return source$.pipe(
+    mergeMap((inner$) =>
+      inner$.pipe(
+        map((v) => [inner$.key, v] as const),
+        endWith([inner$.key, DELETE] as const),
+      ),
     ),
-    defer(() => {
-      res.clear()
-      return of(res)
-    }),
+    scan((acc, [key, value]) => {
+      if (value !== DELETE) return acc.set(key, value)
+      acc.delete(key)
+      return acc
+    }, res),
+    debounceTime(0, asapScheduler),
```

Comment:
Why is this needed? Wouldn't it be better if it was handled by the consumer?

Reply:
I don't know... I've been thinking about this and I'm on the fence.

The thing is that there will be many instances when a single event will trigger many sync changes for different inner-streams. The most obvious one case is when the source stream completes and then all the inner-streams complete one by one, therefore triggering many different deletions until Map is empty.

Of course, that particular case we can handle it in this operator, so that only one empty Map gets emitted when the source completes. But what about the other cases when an event will trigger many different sync changes? Do we really want to emit all those synchronously? Isn't it too noisy?

You are right, though. If we handle the case of the source stream completion, then we can let the consumer do this debouncing themselves... It's just that I can't think of a single case where I would want to get all that noise when what we are emitting is a Map, that's why I think that it makes sense to handle this on operator itself.

Can come up with an example where doing this would be problematic for the consumer?

Reply:
Actually, NVM... I can think of some cases where this can be pretty annoying...